### ISSUE TITLE
write.pcadapt and write.vcf not giving useable results, could this be the POP ID column?

### ISSUE THREAD

**Issue by Ella-Bowles** [thierrygosselin/radiator#28]
Hi Thierry,

I'm trying to do a quick PCAdapt analysis with a tidy file generated in radiator, but it seems that something is going wrong with file conversion to both pcadapt and vcf formats, or that our population labels are not working correctly. Currently, pops are labelled using numbers. I can send you files if it makes it easier to address, but I will write as much as possible here. I am running R v 3.5.1, and did a fresh install of radiator.

When I try to convert the tidy dataframe to pcadapt format using
write_pcadapt("brook_char_tidy_maf.tsv", pop.select = c("1","2","3","4","5","6","7","8","9","10","11","14","16","17"), snp.ld = NULL,
              maf.thresholds = NULL, filename = NULL,
              parallel.core = parallel::detectCores() - 1)

Everything seems to go fine, except that I get a lot of zeros showing up in the data in the console. This seems a little odd.

Small sampling here
$`genotype.matrix`
        Blackfly-d1084 Blackfly-d1191 Blackfly-d1797
   [1,] "0"            "0"            "0"           
   [2,] "9"            "0"            "0"           
   [3,] "9"            "1"            "2"           
        Blackfly-d2016 Blackfly-d2125 Blackfly-d2204
   [1,] "0"            "0"            "0"           
   [2,] "0"            "1"            "1"           
   [3,] "2"            "2"            "2"           
        Blackfly-d2305 Blackfly-d2365 Blackfly-d2388
   [1,] "0"            "0"            "0"           
   [2,] "0"            "1"            "0"           
   [3,] "1"            "0"            "2"           
        Blackfly-d2732 Blackfly-d2771-bf Blackfly-d2975
   [1,] "0"            "0"               "0"           
   [2,] "0"            "0"               "0"           
   [3,] "1"            "2"               "1"           
        Blackfly-d3277 Blackfly-d3469 Blackfly-d3507
   [1,] "0"            "0"            "0"           
   [2,] "0"            "1"            "2"           
   [3,] "2"            "2"            "1"           
        Blackfly-d3580 Blackfly-d3638 Blackfly-d503
   [1,] "0"            "0"            "0"          
   [2,] "0"            "2"            "1"          
   [3,] "2"            "1"            "2"          
        Blackfly-d940 BobsCove-C1028 BobsCove-C1090
   [1,] "0"           "0"            "0"           
   [2,] "0"           "0"            "0"           
   [3,] "0"           "0"            "0"           
        BobsCove-C1140 BobsCove-C1145 BobsCove-C1148
   [1,] "0"            "0"            "0"           
   [2,] "9"            "0"            "0"           
   [3,] "0"            "0"            "0"           
        BobsCove-C1189 BobsCove-C1215 BobsCove-C1401
   [1,] "0"            "0"            "0"           
   [2,] "0"            "0"            "0"           
   [3,] "0"            "0"            "0"           
        BobsCove-C1449 BobsCove-C1476 BobsCove-C1493
   [1,] "0"            "0"            "0"           
   [2,] "0"            "0"            "0"           
   [3,] "0"            "0"            "0"           
        BobsCove-C1526 BobsCove-C1527 BobsCove-C1580
   [1,] "0"            "0"            "0"           
   [2,] "0"            "0"            "0"           
   [3,] "0"            "0"            "0"           
        BobsCove-C1607 BobsCove-C1654 BobsCove-C2255
   [1,] "0"            "0"            "0"           
   [2,] "0"            "0"            "0"           
   [3,] "0"            "0"            "0"           
        BobsCove-C2279 BobsCove-C2304 BobsCove-C2316
   [1,] "0"            "9"            "0"           
   [2,] "0"            "0"            "0"           
   [3,] "0"            "0"            "0"           
        BobsCove-C2390 BobsCove-C2435 BobsCove-C2451
   [1,] "0"            "0"            "0"           
   [2,] "0"            "0"            "0"           
   [3,] "0"            "0"            "0"           
        BobsCove-C2452 BobsCove-C2462 BobsCove-C2530

Looking through the tsv tidy file (which is my collaborators), I see some things that may be a bit odd, but he has produced many analyses successfully with the file, so perhaps it is fine. However, I notice things like that the ALT and REF alleles are the same for the whole file (diff from eachother, but the same for all loci). Anyway, in case this file is fine I run

require(pcadapt)

path_to_file <- "D:/Documents/PostDoc/collaboration-MYates-CapeRace-PondStocking/radiator_pcadapt_20180925@1805.txt"
MY <- read.pcadapt(path_to_file,type="pcadapt")

And I get the error 
1547 lines detected.
324 columns detected.
Warning message:
Only one 'return' characters detected, yet Windows

When I try a different route, using
write_vcf("brook_char_tidy_maf.tsv", pop.info = TRUE, filename = NULL)

I get an error
Error in .f(.x[[i]], ...) : object 'POP_ID' not found

However, I know that this field is in the datafile, so this is weird.

To test to see if the file works without the populations field, I try
write_vcf("brook_char_tidy_maf.tsv", pop.info = FALSE, filename = NULL)

The vcf that I get out of this looks strange in comparison to other vcf files that I have worked with. There are a lot of zeros. Looking at the tidy file though, this may be due to some issue converting the vcf to a tidy file. I can send you both the vcf and the tidy files if you'd like. Anyway, I didn't get any complains from R, so I ran

> require(pcadapt)
> path_to_file <- "D:/Documents/PostDoc/collaboration-MYates-CapeRace-PondStocking/radiator_vcf_file_20180925@1727.vcf"
> MY <- read.pcadapt(path_to_file,type="vcf")

This looks more like what I should see
No variant got discarded.
Summary:

	- input file:				D:/Documents/PostDoc/collaboration-MYates-CapeRace-PondStocking/radiator_vcf_file_20180925@1727.vcf
	- output file:				C:\Users\Admin\AppData\Local\Temp\RtmpCeIy6G\file183c1e3197.pcadapt

	- number of individuals detected:	324
	- number of loci detected:		4614

4614 lines detected.
324 columns detected.

I fetched my output file, and then ran

> x <- pcadapt(MY,K=20)
> plot(x,option="screeplot")

> data <- as.matrix(read.table("file183c1e3197.pcadapt"))
> #check the data
> data[1:5,1:6]
     V1 V2 V3 V4 V5 V6
[1,]  9  9  9  9  9  9
[2,]  9  0  1  2  0  2
[3,]  0  0  0  0  0  0
[4,]  9  9  9  9  9  9
[5,]  9  0  9  0  0  0
> #Without removal of outliers

I skip the population id step b/c we didn't identify them, so go to

> x<-pcadapt(data,K=9)

And then it errors out
Error in UseMethod("pcadapt") : 
  no applicable method for 'pcadapt' applied to an object of class "c('matrix', 'integer', 'numeric')"

Any thoughts about what could be causing these problems? I feel like I should send you my .tsv tidy file, and maybe the vcf and pcadapt files that i generated from those in radiator.

With thanks,
Ella

**Comment by thierrygosselin** [comment]
Sorry for the long delays, got my new computer working today!

I'll go by points/questions/interrogations:

**1. Using the command radiator::write_pcadapt:**

```
write_pcadapt("brook_char_tidy_maf.tsv", pop.select = c("1","2","3","4","5","6","7","8","9","10","11","14","16","17"), snp.ld = NULL,
maf.thresholds = NULL, filename = NULL,
parallel.core = parallel::detectCores() - 1)à
```
You get lots of zeros.

When I run the command with your data I get this:

```
Generating pcadapt file...
pop.select: 
Using markers common in all populations:
    Number of markers before/blacklisted/after:4614/3067/1547
Scanning for monomorphic markers...
    Number of markers before/blacklisted/after: 1547/0/1547
writing pcadapt file with:
    Number of populations: 14
    Number of individuals: 324
    Number of markers: 1547
```
Is this what your getting?

note: pcadapt file is coded 0,1,2,9 and 9 is for missing genotypes.

The only other thing I see here is that your original file : `brook_char_tidy_maf.tsv` as lots of missing data. A lot of the markers are not in common between your groups/sampling sites/pops. This is why in radiator::write_pcadapt you start with 4614 and end up with only 1547. It's pruning the dataset by default because you really don't want to run pcadapt without common markers between your groups.

**Comment by thierrygosselin** [comment]
**2. reading the pcadapt file generated by radiator in pcadapt**

Running this command:
```
require(pcadapt)
test <- pcadapt::read.pcadapt("my_radiator_pca_file", type = "pcadapt")
```

I only get this:

```
1547 lines detected.
324 columns detected.
```

So the warning you're receiving is probably a window issue, I suggest you check the file in a pc text editor to see if you have something strange.

**Comment by thierrygosselin** [comment]
**3. Using radiator::write_vcf**

```
check <- radiator::write_vcf("brook_char_tidy_maf.tsv", pop.info = TRUE, filename = NULL)
```

I get:

```
 Error in .f(.x[[i]], ...) : object 'POP_ID' not found 
```

definitely a bug.


However, try running with `pop.info = FALSE` like this:
```
check <- radiator::write_vcf("brook_char_tidy_maf.tsv", filename = NULL)
```

Here's the doc concerning this argument:

```
_pop.info | (optional, logical)_ 
Should the population information be included in the FORMAT field
(along the GT info for each samples ?). 
To make the VCF population-ready use pop.info = TRUE. 
The population information must be included in the POP_ID column of the tidy dataset.
Default: pop.info = FALSE.
```

It's not a fully functional arguments yet, no other software besides radiator is exploiting the pop info in vcf. So there's no harm in turning it off as it is by default...

**Comment by thierrygosselin** [comment]
**4. Using the vcf file as pcadapt input file**

```
require(pcadapt)
test <- pcadapt::read.pcadapt(path_to_file,type="vcf")
```

I get:

```
No variant got discarded.
Summary:

	- input file:				~radiator_vcf_file_20181001@1634.vcf
	- output file:				~filea31cda59f03.pcadapt

	- number of individuals detected:	324
	- number of loci detected:		4614

4614 lines detected.
324 columns detected.
```

As you say above no markers got discarded here. And it's normal, because I havent applied a default _common.markers = TRUE_ filtering when the vcf was generated in radiator.

**Comment by thierrygosselin** [comment]
**5. problem while running your file in pcadapt**

```
x <- pcadapt::pcadapt(data, K = 9)
```

You get this error:
```
Error in UseMethod("pcadapt") :
no applicable method for 'pcadapt' applied to an object of class "c('matrix', 'integer', 'numeric')"
```

I don't have an error:

```
names(x)
[1] "scores"          "singular.values" "loadings"        "zscores"         "af"              "maf"            
 [7] "chi2.stat"       "stat"            "gif"             "pvalues"         "pass"    
```
At this point here it's a pcadapt problem, try re-installing it. If nothing work, ask Michael Blum for help.
I've send you the result for the analysis with and without common markers...

Cheers
Thierry

**Comment by Ella-Bowles** [comment]
Thanks Thierry,

I was able to get a good chunk of the results that you were as well.

I did check the file generated at "test <- pcadapt::read.pcadapt("my_radiator_pca_file", type = "pcadapt")"  for windows line endings, and my machine is telling me that it is only ASCII, so not sure what is happening here.

I can get the same as you get from "x <- pcadapt::pcadapt(data, K = 9) and then names(x)" if I use the original dataset generated as "test " in read.pcadapt above, but not if I convert to a matrix (using data <- as.matrix(read.table("file32742f854a4d.pcadapt.bed")), and then use the "data" object. I see many errors when I try to convert to the matrix. Given the error that I am getting with the read.pcadapt command though, I suspect that there is something wrong with the output bed file. Anyway, I am awaiting Mike Blum's feedback on my errors at the moment, and I'll let you know if things don't work.

As an aside, the "common.markers = TRUE" argument isn't working for me when I try to generate the vcf file in radiator. Not a big deal if I can get the pcadapt file to work, but just wanted to give you a heads up.

Ella
